controllers/prometheus_exporter.go

```diff
@@ -0,0 +1,202 @@
+    mdbv1 "github.com/mongodb/mongodb-kubernetes-operator/api/v1"
+    corev1 "k8s.io/api/core/v1"
+)
+
+const (
+    exporterUserName = "metrics"
+    clusterDomain = "cluster.local"
+    prometheusNamespace = "monitoring"
+)
+
+// buildExporterServiceAccount creates a ServiceAccount that will be used for the Prometheus Exporter
+func buildExporterServiceAccount(mdb mdbv1.MongoDBCommunity) corev1.ServiceAccount {
+    return corev1.ServiceAccount {
+        ObjectMeta: metav1.ObjectMeta {
+            Name: mdb.Name() + "-exporter"
+            Namespace: mdb.Namespace()
+            Labels: struct {
+                "app": mdb.Name() + "-exporter"
+            }
+        }
+    }
+}
+
+// buildMongoDbUriSecret creates a Secret with the URI of the MongoDB instance to monitor
+func buildMongoDbUriSecret(mdb mdbv1.MongoDBCommunity) corev1.Secret {
+    label := make(map[string]string)
+    label["app"] = mdb.Name() + '-exporter'
+
+    password := ''
+    for _, user := range mdb.Users {
+        if user.Name() == exporterUserName {
+            secretsClient := r.client.CoreV1().Secrets(mdb.Namespace)
+            passwordRef := corev1.Secret{}
+            err := secretsClient.Get(context.TODO(), user.PasswordSecretRef.Name(), %passwordRef)
+            password = passwordRef.Data[user.PasswordSecretRef.Key]
+            break
+        }
+    }
+
+    full_uri = fmt.Sprintf(
+        "mongodb://%s:%s@%s.%s.%s.%s:%d",
+        exporterUserName,
+        password,
+        m.ServiceName(),
+        m.Namespace,
+        clusterDomain,
+        27017)
+
+    return secret.Builder().
+        SetName(mdb.Name() + '-exporter-uri').
+        SetNamespace(mdb.Namespace).
+        SetLabels(label).
+        SetField('mongodb-uri', full_uri).
+        Build()
+}
+
+func (r *ReplicaSetReconciler) ensureExporterDeployment(mdb mdbv1.MongoDBCommunity) error {
+    dpl := buildExporterDeployment(mdb)
+    err := r.client.Create(context.TODO(), &dpl)
+    if err != nil && apiErrors.IsAlreadyExists(err) {
+        r.log.Infof("The deployment already exists... moving forward: %s", err)
+        return nil
+    }
+    return err
+}
+
+// buildExporterDeployment creates a Deployment for the Prometheus Exporter
+func buildExporterDeployment(mdb mdbv1.MongoDBCommunity) appsv1.Deployment {
+    label := make(map[string]string)
+    label["app"] = mdb.Name() + '-exporter'
+
+    ExporterServiceAccount := buildExporterServiceAccount(mdb)
+    err := r.client.Create(context.TODO(), &ExporterServiceAccount)
+    if err != nil && !apiErrors.IsAlreadyExists(err) {
+        r.log.Infof("Error when building exporter service account: %s", err)
+        return err
+    }
+
+    MongoDbUriSecret := buildMongoDbUriSecret(mdb)
+    err := r.client.Create(context.TODO(), &MongoDbUriSecret)
+    if err != nil && !apiErrors.IsAlreadyExists(err) {
+        r.log.Infof("Error when building Mongo DB URI Secret: %s", err)
+        return err
+    }
+
+    return deployment.Builder().
```

Comment:
This is neat, I have never seen the builder pattern before in golang. Generally people just pass structs around instead of doing this kind of pattern.